app/Models/Extensions/HasTimeBasedID.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace App\Models\Extensions;
+
+use App\Models\Configs;
+use App\Models\Logs;
+use Illuminate\Database\Eloquent\Concerns\HasAttributes;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\QueryException;
+
+/**
+ * Trait HasTimeBasedID.
+ *
+ * Inspired by https://emymbenoun.medium.com/how-to-use-uuids-instead-of-auto-increment-ids-in-your-laravel-app-2e6cc045f6c1.
+ */
+trait HasTimeBasedID
+{
+	public static function bootHasTimeBasedID()
+	{
+		static::creating(function (Model $model) {
+			$pKey = $model->getKeyName();
+			if (empty($model->$pKey)) {
+				$model->generateID();
+			}
+
+			return true;
+		});
+	}
+
+	/**
+	 * Get the value indicating whether the IDs are incrementing.
+	 *
+	 * @return bool
+	 */
+	public function getIncrementing(): bool
+	{
+		return false;
+	}
+
+	/**
+	 * Set whether IDs are incrementing.
+	 *
+	 * @param bool $value
+	 */
+	public function setIncrementing($value)
+	{
+		throw new \BadMethodCallException('must not call setIncrementing for a model which uses the trait HasTimeBasedID');
+	}
+
+	/**
+	 * Set a given attribute on the model.
+	 *
+	 * Actually, this method should override
+	 * {@link HasAttributes::setAttribute()} in order to prevent setting of
+	 * the primary key by application code.
+	 * However, {@link AlbumCast::toTagAlbum()} does it nonetheless, so we
+	 * don't throw an exception until that method is fixed.
+	 *
+	 * TODO: Fix {@link AlbumCast::toTagAlbum()}.
+	 *
+	 * @param string $key   name of attribute which is being set
+	 * @param mixed  $value value of attribute
+	 *
+	 * @return mixed
+	 */
+	public function setAttribute($key, $value)
+	{
+		// TODO: Uncomment the following lines, after AlbumCast::toTagAlbum() has been fixed
+		/*if ($key == $this->getKeyName()) {
+			throw new \InvalidArgumentException('must not set primary key explicitly, primary key will be set on first insert');
+		}*/
+
+		return parent::setAttribute($key, $value);
+	}
+
+	/**
+	 * Saves the model to the database.
+	 *
+	 * @param array $options
+	 *
+	 * @return bool
+	 */
+	public function save(array $options = []): bool
+	{
+		$result = false;
+		$retryCounter = 5;
```

Comment:
This limit is a new thing, isn't it? On the one hand, I agree that looping _forever_ was a poor style. On the other hand, is `5` large enough? Keep in mind that for users with 32-bit IDs, the IDs have a full-second resolution, so they are _very_ likely to conflict when uploading multiple photos at a time...

Reply:
Yes, the limit is a new thing to avoid infinite loops. I expect five to be sufficiently large. The IDs are time-based, i.e. strictly monotonic and not random (so the birthday paradox does not apply).

Moreover, PHP is single-threaded within the _same_ request. Hence, if a user uploads or imports a set of photos within a single request then we have no problem even if the user as a high-speed connection and several photos are imported during the same second. If PHP sleeps for one second, then the whole sequence of photos is delayed. (In this case, the limit 1 should actually be enough.)

We only run into problems, if multiple users (or the same user) uploads photos _parallel_ requests. Honestly, I have no gut feeling which would be an appropriate limit in this case.

Reply:
> Moreover, PHP is single-threaded within the same request. Hence, if a user uploads or imports a set of photos within a single request then we have no problem even if the user as a high-speed connection and several photos are imported during the same second. If PHP sleeps for one second, then the whole sequence of photos is delayed. (In this case, the limit 1 should actually be enough.)

Actually no, the JS front-end throws multiple requests at the same time to the server, that is why you are sometimes processing 4 images in parallel (and it is a good thing as long as you don't hit a memory limit).

Though I do expect the limit of 5 to be sufficiently large. :)